Feature/swipe function

IsabellUrban opened the pull request.

onemanwenttomow commented on components/Flashcard/Flashcard.js:
```diff
@@ -32,6 +39,41 @@ export default function Flashcard({
+  );
+  const animControls = useAnimation();
+
+  function handleDragEnd() {
+    (event, info) => {
+      const { offset } = info;
+
+      if (!showAnswer) return;
+
+      if (offset.x > 150) {
+        onSwipe("right");
+        animControls.start({ x: 400, opacity: 0 }).then(() => {
+          motionValue.set(0);
+          animControls.start({ x: 0, opacity: 1 });
+        });
+      } else if (offset.x < -50) {
+        onSwipe("left");
+        animControls.start({ x: -400, opacity: 0 }).then(() => {
+          motionValue.set(0);
+          animControls.start({ x: 0, opacity: 1 });
+        });
+      } else {
+        animControls.start({ x: 0, opacity: 1 });
+      }
+    };
+  }
```
I think this function also needs a little work. 

We can get the distance of the drag from the `motionValue` object.
We no longer need to worry about resetting the animation if you set the value of the dragConstraints to 0. 
(This also means we can remove the final else). 
I optionally put in that the card should be re-flipped at the end of the drag event (but you can decide if you want this or not, or when it should happen. 
Maybe something like:

``` js
  function handleDragEnd() {
    if (!showAnswer) return;
    const x = motionValue.get();

    console.log(x);

    if (x > 30) {
      console.log("increase");
      onSwipe("right");
      onIncreaseFlashcardLevel(id);
    } else if (x < -30) {
      console.log("decrease");
      onSwipe("left");
      onDecreaseFlashcardLevel(id);
    }
    setShowAnswer(false);
  }

```

onemanwenttomow reviewed (changes requested):
Hey team. Managed to find a little bit of time to take a look. Added some thoughts and suggestions (but they are by no means definitive).

IsabellUrban commented on package.json:
```diff
@@ -13,6 +13,8 @@
     "color-contrast-finder": "^1.0.2",
     "eslint": "8.44.0",
     "eslint-config-next": "13.4.8",
+    "framer": "^2.4.1",
```
Thanks, I uninstalled framer and now it works.

IsabellUrban commented on pages/training/[id].js:
```diff
@@ -14,10 +15,19 @@ export default function CollectionPage({
 
-  const collectionTitle = allFlashcardsFromCollection?.[0]?.collectionTitle;
-  const collectionColor = allFlashcardsFromCollection?.[0]?.collectionColor;
+  useEffect(() => {
+    if (id) {
+      const allFlashcardsFromCollection = getAllFlashcardsFromCollection(id);
+      setSortedFlashcards(sortFlashcardsByLevel(allFlashcardsFromCollection));
+    }
+  }, [id, getAllFlashcardsFromCollection]);
```
If I don't put in an useEffect Hook I get an error message saying that there are too many re-renders. React limits the number of renders to prevent an infinite loop. I put in the early return so when there is no id it stops but I left the useEffect in.

IsabellUrban commented:
Thanks Pete for your detailed comment and the ideas to fix it. Now everything finally works.

onemanwenttomow commented:
Great work so far team. I think i would attempt to simplify things though. 

1. Rather than showing the next 3 cards, i would map over all the cards and stack them on top of each other with grid. If you are worried about stacking too many cards you could always slice the sorted array. 
2. I would remove the icon idea for now. If you want to do this i think you need to position this item somewhere else in the dom (if it is meant to be a toast notification).
3. We need to add `overflow: hidden` to the main. To prevent the page resizing once the card is dragged and then needs to flip again. 
4. Remove the function call `onSwipe` as i feel like this is doing the work for a second time. You already do it with `onIncreaseFlashcardLevel` inside of `handleDragEnd`
5. Have a reset function that sets the state back to the original after the drag event. 

``` js
// [id].js 

import { useRouter } from "next/router";
import { useState, useEffect } from "react";
import styled from "styled-components";
import Flashcard from "@/components/Flashcard/Flashcard";

export default function CollectionPage({
  handleToggleCorrect,
  handleDeleteFlashcard,
  changeActionMode,
  changeCurrentFlashcard,
  getAllFlashcardsFromCollection,
  handleIncreaseFlashcardLevel,
  handleDecreaseFlashcardLevel,
  handleFirstClick,
}) {
  const router = useRouter();
  const { id } = router.query;

  const [sortedFlashcards, setSortedFlashcards] = useState([]);

  useEffect(() => {
    if (!id) return;
    const allFlashcardsFromCollection = getAllFlashcardsFromCollection(id);
    setSortedFlashcards(sortFlashcardsByLevel(allFlashcardsFromCollection));
  }, [id, getAllFlashcardsFromCollection]);

  function sortFlashcardsByLevel(flashcards) {
    return flashcards.sort((a, b) => {
      if (a.level !== b.level) {
        return a.level - b.level;
      }

      if (!a.trainingDate && !b.trainingDate) {
        return 0;
      } else if (!a.trainingDate) {
        return -1;
      } else if (!b.trainingDate) {
        return 1;
      } else {
        return a.trainingDate - b.trainingDate;
      }
    });
  }

  function areAllCardsLearned(flashcards) {
    return flashcards.length > 0 && flashcards[0].level === 5;
  }

  const allLearned = areAllCardsLearned(sortedFlashcards);

  const currentFlashcard = sortedFlashcards[0];
  const nextFlashcard = sortedFlashcards[1];
  const thirdFlashcard = sortedFlashcards[2];

  function handleSwipe(direction) {
    if (direction === "right") {
      handleIncreaseFlashcardLevel(currentFlashcard.id);
    } else if (direction === "left") {
      handleDecreaseFlashcardLevel(currentFlashcard.id);
    }
  }

  return (
    <>
      {sortedFlashcards.length > 0 && (
        <>
          <StyledHeadline>
            {sortedFlashcards[0]?.collectionTitle}
          </StyledHeadline>
          <StyledSubheading>Training Mode</StyledSubheading>

          {allLearned && (
            <StyledSuccessMessage>
              You have trained all flashcards up to the last level, BrainStack
              Champion! <span aria-label="party-popper-emoji">🎉</span>
            </StyledSuccessMessage>
          )}

          <FlashcardListWrapper>
            {/* <FlashcardStackWrapper> */}
            {/* <ThirdFlashcardWrapper>
                <Flashcard
                  collectionColor={currentFlashcard?.collectionColor}
                  handleDeleteFlashcard={handleDeleteFlashcard}
                  flashcard={thirdFlashcard}
                  onToggleCorrect={handleToggleCorrect}
                  changeCurrentFlashcard={changeCurrentFlashcard}
                  changeActionMode={changeActionMode}
                  modeSelection="training"
                  onIncreaseFlashcardLevel={handleIncreaseFlashcardLevel}
                  onDecreaseFlashcardLevel={handleDecreaseFlashcardLevel}
                  handleFirstClick={handleFirstClick}
                />
              </ThirdFlashcardWrapper>
              <ThirdFlashcardOpacity />
              <SecondFlashcardWrapper>
                <Flashcard
                  collectionColor={currentFlashcard?.collectionColor}
                  handleDeleteFlashcard={handleDeleteFlashcard}
                  flashcard={nextFlashcard}
                  onToggleCorrect={handleToggleCorrect}
                  changeCurrentFlashcard={changeCurrentFlashcard}
                  changeActionMode={changeActionMode}
                  modeSelection="training"
                  onIncreaseFlashcardLevel={handleIncreaseFlashcardLevel}
                  onDecreaseFlashcardLevel={handleDecreaseFlashcardLevel}
                  handleFirstClick={handleFirstClick}
                />
              </SecondFlashcardWrapper>
              <SecondFlashcardOpacity /> */}
            {sortedFlashcards.toReversed().map((card) => (
              <Flashcard
                key={card.id}
                collectionColor={card?.collectionColor}
                handleDeleteFlashcard={handleDeleteFlashcard}
                flashcard={card}
                onToggleCorrect={handleToggleCorrect}
                changeCurrentFlashcard={changeCurrentFlashcard}
                changeActionMode={changeActionMode}
                modeSelection="training"
                onIncreaseFlashcardLevel={handleIncreaseFlashcardLevel}
                onDecreaseFlashcardLevel={handleDecreaseFlashcardLevel}
                handleFirstClick={handleFirstClick}
              />
            ))}
            {/* </FlashcardStackWrapper> */}
          </FlashcardListWrapper>
        </>
      )}
      {(!sortedFlashcards || sortedFlashcards.length === 0) && (
        <StyledMessage>
          No flashcards in this collection found. Add some new!
        </StyledMessage>
      )}
    </>
  );
}

const StyledMessage = styled.p`
  text-align: center;
  font-size: 1rem;
  padding: 40px 20px;
`;

const FlashcardListWrapper = styled.ul`
  display: grid;
  /* align-items: center;
  flex-direction: column; */
  /* gap: 15px; */
  list-style: none;
`;

const StyledHeadline = styled.h2`
  font: var(--main-headline);
  text-align: center;
  padding-top: 35px;
`;

const StyledSubheading = styled.h3`
  font: var(--sub-headline);
  text-align: center;
  padding: 5px 0 30px 0;
`;

const StyledSuccessMessage = styled.p`
  font: var(--main-headline);
  text-align: center;
  color: var(--primary-green);
  padding: 0 30px 50px 30px;
`;

const FlashcardStackWrapper = styled.div`
  margin: 0 auto;
  width: 90vw;
  height: 218px;
  max-width: 500px;
  position: relative;

  @media (min-width: 768px) {
    min-height: 300px;
    max-height: 400px;
  }
`;

const FlashcardStackShadow = styled.div`
  margin: 0 auto;
  width: 90vw;
  height: 218px;
  max-width: 500px;
  position: absolute;
  border-radius: 10px;

  @media (min-width: 768px) {
    min-height: 300px;
    max-height: 400px;
  }
`;

const SecondFlashcardWrapper = styled(FlashcardStackShadow)`
  top: 8px;
  left: 8px;
  background-color: var(--secondary-light-grey);
`;

const ThirdFlashcardWrapper = styled(FlashcardStackShadow)`
  top: 16px;
  left: -8px;
  background-color: var(--secondary-grey);
`;

const SecondFlashcardOpacity = styled(FlashcardStackShadow)`
  top: 8px;
  left: 8px;
  background-color: var(--secondary-light-grey);
  opacity: 0.6;
`;

const ThirdFlashcardOpacity = styled(FlashcardStackShadow)`
  top: 16px;
  left: -8px;
  background-color: var(--secondary-grey);
  opacity: 0.6;
`;


```


``` js
// Flashcard.js (removed the useEffects and tried to simplify things a little


import styled from "styled-components";
import { useState } from "react";
import { findContrastColor } from "color-contrast-finder";
import { motion, useMotionValue, useTransform } from "framer-motion";
import MarkAsIncorrect from "@/public/icons/MarkAsIncorrect.svg";
import MarkAsCorrect from "@/public/icons/MarkAsCorrect.svg";
import Delete from "@/public/icons/Delete.svg";
import Edit from "@/public/icons/Edit.svg";
import RoundButton from "../Buttons/RoundButton";
import DeleteConfirmationDialog from "../DeleteConfirmationDialog/DeleteConfirmationDialog";
import LevelBar from "../LevelBar/LevelBar";

export default function Flashcard({
  flashcard,
  handleDeleteFlashcard,
  changeCurrentFlashcard,
  changeActionMode,
  collectionColor,
  modeSelection,
  onIncreaseFlashcardLevel,
  onDecreaseFlashcardLevel,
  onToggleCorrect,
  handleFirstClick,
}) {
  const [showAnswer, setShowAnswer] = useState(false);
  const [isDelete, setIsDelete] = useState(false);
  const [isVisible, setIsVisible] = useState(true);

  const {
    question,
    answer,
    collectionTitle: collection,
    id,
    isCorrect,
    level,
  } = flashcard;

  const motionValue = useMotionValue(0);
  const rotateValue = useTransform(motionValue, [-300, 300], [-100, 100]);
  const opacityValue = useTransform(
    motionValue,
    [-150, -75, 0, 75, 150],
    [0, 0, 1, 0, 0]
  );

  function resetState() {
    setShowAnswer(false);
    setIsDelete(false);
    setIsVisible(true);
  }

  function handleDragEnd() {
    if (!showAnswer) return;
    const x = motionValue.get();

    if (x > 30) {
      onIncreaseFlashcardLevel(id);
      resetState();
    } else if (x < -30) {
      onDecreaseFlashcardLevel(id);
      resetState();
    }
  }

  function handleShowAnswer() {
    setShowAnswer(!showAnswer);
  }

  function toggleDeleteConfirmation(event) {
    event.stopPropagation();
    setIsDelete(!isDelete);
  }

  function setEditWithoutFlip(event) {
    event.stopPropagation();
    changeActionMode("edit");
    changeCurrentFlashcard(flashcard);
  }

  function handleMarkAsCorrect(id) {
    handleFirstClick();
    onIncreaseFlashcardLevel(id);
  }

  function handleMarkAsInCorrect(id) {
    handleFirstClick();
    onDecreaseFlashcardLevel(id);
  }

  const contrastOptions = {
    color: collectionColor,
    threshold: 0.5,
    highColor: "#000",
    lowColor: "#fff",
  };

  const titleColor = findContrastColor(contrastOptions);

  return (
    <>
      {isVisible && modeSelection === "training" ? (
        <motion.div
          drag={showAnswer ? "x" : false}
          dragConstraints={{ left: 0, right: 0 }}
          style={{
            x: motionValue,
            rotate: rotateValue,
            opacity: opacityValue,
            gridRow: "1 / 1",
            gridColumn: "1 / 1",
          }}
          onDragEnd={handleDragEnd}
          transition={{ type: "spring", duration: 0.5 }}
        >
          <CardContainer onClick={handleShowAnswer}>
            <StyledFlashcard $showAnswer={showAnswer}>
              {isDelete && (
                <>
                  <CardFront $collectionColor={collectionColor}>
                    <StyledDeleteConfirmationDialogContainer>
                      <DeleteConfirmationDialog
                        onDeleteFlashcard={handleDeleteFlashcard}
                        toggleDeleteConfirmation={toggleDeleteConfirmation}
                        id={id}
                        variant="flashcard"
                      />
                    </StyledDeleteConfirmationDialogContainer>
                  </CardFront>
                  <CardBack $collectionColor={collectionColor}>
                    <StyledDeleteConfirmationDialogContainer>
                      <DeleteConfirmationDialog
                        onDeleteFlashcard={handleDeleteFlashcard}
                        toggleDeleteConfirmation={toggleDeleteConfirmation}
                        id={id}
                        variant="flashcard"
                      />
                    </StyledDeleteConfirmationDialogContainer>
                  </CardBack>
                </>
              )}

              {!isDelete && (
                <>
                  <CardFront $collectionColor={collectionColor}>
                    <CollectionTitle
                      $collectionColor={collectionColor}
                      $titleColor={titleColor}
                    >
                      {collection}
                    </CollectionTitle>
                    <StyledEditButtonContainer>
                      <RoundButton
                        onClick={setEditWithoutFlip}
                        type="button"
                        variant="edit"
                      >
                        <Edit />
                      </RoundButton>
                    </StyledEditButtonContainer>
                    <StyledDeleteButtonContainer>
                      <RoundButton
                        onClick={toggleDeleteConfirmation}
                        type="button"
                        variant="delete"
                      >
                        <Delete />
                      </RoundButton>
                    </StyledDeleteButtonContainer>

                    <Question>{question}</Question>

                    {modeSelection === "training" && (
                      <LevelBarWrapper>
                        <LevelBar level={level} />
                      </LevelBarWrapper>
                    )}

                    {modeSelection === "learning" && isCorrect && (
                      <StyledCorrectIcon>
                        <MarkAsCorrect />
                      </StyledCorrectIcon>
                    )}
                  </CardFront>

                  <CardBack $collectionColor={collectionColor}>
                    <StyledDeleteButtonContainer>
                      <RoundButton
                        onClick={toggleDeleteConfirmation}
                        type="button"
                        variant="delete"
                      >
                        <Delete />
                      </RoundButton>
                    </StyledDeleteButtonContainer>
                    <Answer>{answer}</Answer>

                    {modeSelection === "learning" && (
                      <StyledMarkAsButtonContainer>
                        <RoundButton
                          onClick={() => onToggleCorrect(id)}
                          type="button"
                          variant={
                            isCorrect ? "markAsIncorrect" : "markAsCorrect"
                          }
                        >
                          {isCorrect ? <MarkAsIncorrect /> : <MarkAsCorrect />}
                        </RoundButton>
                      </StyledMarkAsButtonContainer>
                    )}

                    {modeSelection === "training" && (
                      <>
                        <StyledIncorrectButtonContainer>
                          <RoundButton
                            onClick={() => {
                              handleMarkAsInCorrect(id);
                            }}
                            type="button"
                            variant={"markAsIncorrect"}
                          >
                            <MarkAsIncorrect />
                          </RoundButton>
                        </StyledIncorrectButtonContainer>
                        <StyledCorrectButtonContainer>
                          <RoundButton
                            onClick={() => {
                              handleMarkAsCorrect(id);
                            }}
                            type="button"
                            variant={"markAsCorrect"}
                          >
                            <MarkAsCorrect />
                          </RoundButton>
                        </StyledCorrectButtonContainer>
                      </>
                    )}
                  </CardBack>
                </>
              )}
            </StyledFlashcard>
          </CardContainer>
        </motion.div>
      ) : (
        <CardContainer onClick={handleShowAnswer}>
          <StyledFlashcard $showAnswer={showAnswer}>
            {isDelete && (
              <>
                <CardFront $collectionColor={collectionColor}>
                  <StyledDeleteConfirmationDialogContainer>
                    <DeleteConfirmationDialog
                      onDeleteFlashcard={handleDeleteFlashcard}
                      toggleDeleteConfirmation={toggleDeleteConfirmation}
                      id={id}
                      variant="flashcard"
                    />
                  </StyledDeleteConfirmationDialogContainer>
                </CardFront>
                <CardBack $collectionColor={collectionColor}>
                  <StyledDeleteConfirmationDialogContainer>
                    <DeleteConfirmationDialog
                      onDeleteFlashcard={handleDeleteFlashcard}
                      toggleDeleteConfirmation={toggleDeleteConfirmation}
                      id={id}
                      variant="flashcard"
                    />
                  </StyledDeleteConfirmationDialogContainer>
                </CardBack>
              </>
            )}

            {!isDelete && (
              <>
                <CardFront $collectionColor={collectionColor}>
                  <CollectionTitle
                    $collectionColor={collectionColor}
                    $titleColor={titleColor}
                  >
                    {collection}
                  </CollectionTitle>
                  <StyledEditButtonContainer>
                    <RoundButton
                      onClick={setEditWithoutFlip}
                      type="button"
                      variant="edit"
                    >
                      <Edit />
                    </RoundButton>
                  </StyledEditButtonContainer>
                  <StyledDeleteButtonContainer>
                    <RoundButton
                      onClick={toggleDeleteConfirmation}
                      type="button"
                      variant="delete"
                    >
                      <Delete />
                    </RoundButton>
                  </StyledDeleteButtonContainer>

                  <Question>{question}</Question>

                  {modeSelection === "training" && (
                    <LevelBarWrapper>
                      <LevelBar level={level} />
                    </LevelBarWrapper>
                  )}

                  {modeSelection === "learning" && isCorrect && (
                    <StyledCorrectIcon>
                      <MarkAsCorrect />
                    </StyledCorrectIcon>
                  )}
                </CardFront>

                <CardBack $collectionColor={collectionColor}>
                  <StyledDeleteButtonContainer>
                    <RoundButton
                      onClick={toggleDeleteConfirmation}
                      type="button"
                      variant="delete"
                    >
                      <Delete />
                    </RoundButton>
                  </StyledDeleteButtonContainer>
                  <Answer>{answer}</Answer>

                  {modeSelection === "learning" && (
                    <StyledMarkAsButtonContainer>
                      <RoundButton
                        onClick={() => onToggleCorrect(id)}
                        type="button"
                        variant={
                          isCorrect ? "markAsIncorrect" : "markAsCorrect"
                        }
                      >
                        {isCorrect ? <MarkAsIncorrect /> : <MarkAsCorrect />}
                      </RoundButton>
                    </StyledMarkAsButtonContainer>
                  )}

                  {modeSelection === "training" && (
                    <>
                      <StyledIncorrectButtonContainer>
                        <RoundButton
                          onClick={() => onDecreaseFlashcardLevel(id)}
                          type="button"
                          variant={"markAsIncorrect"}
                        >
                          <MarkAsIncorrect />
                        </RoundButton>
                      </StyledIncorrectButtonContainer>
                      <StyledCorrectButtonContainer>
                        <RoundButton
                          onClick={() => onIncreaseFlashcardLevel(id)}
                          type="button"
                          variant={"markAsCorrect"}
                        >
                          <MarkAsCorrect />
                        </RoundButton>
                      </StyledCorrectButtonContainer>
                    </>
                  )}
                </CardBack>
              </>
            )}
          </StyledFlashcard>
        </CardContainer>
      )}
    </>
  );
}

const CardContainer = styled.li`
  perspective: 1000px;
`;

const StyledFlashcard = styled.article`
  width: 90vw;
  height: auto;
  min-height: 218px;
  max-width: 500px;
  position: relative;
  transform-style: preserve-3d;
  transition: transform 0.3s, transform 0.6s;
  transform: ${({ $showAnswer }) =>
    $showAnswer ? "rotateY(180deg)" : "rotateY(0deg)"};

  cursor: pointer;

  @media (min-width: 768px) {
    min-height: 300px;
    transition: transform 0.8s;
  }
`;

const CardFace = styled.div`
  display: grid;
  grid-template-columns: var(--grid-columns-card-and-title);
  grid-template-rows: var(--grid-rows-card-and-title);
  position: absolute;
  width: 100%;
  height: 100%;
  backface-visibility: hidden;
  border-radius: 10px;
  border: var(--border-thickness) solid
    ${({ $collectionColor }) => $collectionColor};
`;

const CardFront = styled(CardFace)`
  background-color: #fff;
`;

const CardBack = styled(CardFace)`
  background-color: var(--secondary-light-grey);
  transform: rotateY(180deg);
`;

const CollectionTitle = styled.p`
  font: var(--collection-title);
  display: inline-block;
  color: ${({ $titleColor }) => $titleColor};
  background-color: ${({ $collectionColor }) => $collectionColor};
  padding: 5px 12px 7px 10px;
  grid-column: 1 / 6;
  grid-row: 1 / 2;
  justify-self: start;
  align-self: start;
  border-top-left-radius: 6px;
  border-bottom-right-radius: 10px;
`;

const Answer = styled.p`
  font: var(--question-answer);
  padding: 10px;
  grid-column: 1 / 8;
  grid-row: 2 / 3;
  align-self: center;

  @media (min-width: 768px) {
    font-size: 1.4rem;
  }
`;

const Question = styled.h3`
  font: var(--question-answer);
  padding: 10px;
  white-space: normal;
  grid-column: 1 / 8;
  grid-row: 2 / 3;
  align-self: center;

  @media (min-width: 768px) {
    font-size: 1.4rem;
  }
`;

const StyledCorrectIcon = styled.div`
  color: var(--primary-green);
  display: flex;
  align-items: center;
  justify-content: center;
  grid-column: 7 / 8;
  grid-row: 3 / 4;
`;

const RoundButtonContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const StyledEditButtonContainer = styled(RoundButtonContainer)`
  grid-column: 6 / 7;
  grid-row: 1 / 2;
`;

const StyledDeleteButtonContainer = styled(RoundButtonContainer)`
  grid-column: 7 / 8;
  grid-row: 1 / 2;
`;

const StyledMarkAsButtonContainer = styled(RoundButtonContainer)`
  grid-column: 7 / 8;
  grid-row: 3 / 4;
`;

const StyledIncorrectButtonContainer = styled(RoundButtonContainer)`
  grid-column: 6 / 7;
  grid-row: 3 / 4;
`;

const StyledCorrectButtonContainer = styled(RoundButtonContainer)`
  grid-column: 7 / 8;
  grid-row: 3 / 4;
`;

const StyledDeleteConfirmationDialogContainer = styled.div`
  grid-column: 1 / 8;
  grid-row: 2 / 4;

  @media (min-width: 768px) {
    padding-top: 40px;
  }
`;

const LevelBarWrapper = styled.div`
  padding-top: 20px;
  grid-column: 1 / 8;
  grid-row: 3 / 4;
`;


```

IsabellUrban commented:
Hi Pete, thanks again for your help. I have added no. 1 as you suggested. To get the card deck design back I took the design with the shadows from the collection. I moved the icon idea to a new component called bubble animation and added it to the flashcard (it is currently commented out in the flashcard). Unfortunately it doesn't work and I don't know why. Adding overflow hidden to the main didn't have the desired effect (you couldn't scroll vertically anymore). I have added an extra container in the [id].js with overflow: hidden that works. I added no. 5 as suggested. Thanks a lot. :-)

Translated with DeepL.com (free version)

ahohnsen reviewed (approved):
Well done! 🥳 I am glad you finally finished that feature. 👏🏼 

Please, have a look at my comments. They are mostly regarding refactoring suggestions and best practices. Please, be consistent with styling and create styled components for your `motion.div`. 🙂

IsabellUrban commented on components/BubbleAnimation/BubbleAnimation.js:
```diff
@@ -0,0 +1,48 @@
+export default function BubbleAnimation({ swipeDirection }) {
+  if (!swipeDirection) return null;
+
+  const SwipeIcon = ({ direction }) => (
+    <motion.div
+      initial={{ opacity: 0, scale: 0.8 }}
+      animate={{
+        scale: [1, 1.1, 1],
+        opacity: [0.0, 0.7, 0.5, 0],
+      }}
+      transition={{ duration: 1.5, ease: "easeOut" }}
+      style={{
+        position: "absolute",
+        top: "-65px",
+        left: direction === "left" ? "2%" : "auto",
+        right: direction === "right" ? "2%" : "auto",
+        backgroundColor:
+          direction === "left" ? "var(--primary-red)" : "var(--primary-green)",
+        borderRadius: "50%",
+        width: "40px",
+        height: "40px",
+        color: "#fff",
+        display: "flex",
+        justifyContent: "center",
+        alignItems: "center",
+        boxShadow: "0px 0px 7.5px rgba(0, 0, 0, 0.3)",
+        zIndex: 1000,
+      }}
+    >
+      {direction === "left" ? (
+        <MarkAsIncorrect style={{ color: "white" }} />
+      ) : (
+        <MarkAsCorrect style={{ color: "white" }} />
```
Thanks, I tried to make a styled component out of it with `styled.motion.div` which did not work. I then thought that maybe you can't make styled components from `motion.div` at all and discarded it for the time being. Of course, it's much clearer and makes more sense this way.

IsabellUrban commented on components/Flashcard/Flashcard.js:
```diff
@@ -57,127 +121,276 @@ export default function Flashcard({
+    <>
+      {isAnimating && <BubbleAnimation swipeDirection={swipeDirection} />}
+      {isVisible && modeSelection === "training" ? (
+        <motion.div
+          drag={showAnswer ? "x" : false}
+          dragConstraints={{ left: 0, right: 0 }}
+          style={{
+            x: motionValue,
+            rotate: rotateValue,
+            opacity: opacityValue,
+            gridRow: "1 / 1",
+            gridColumn: "1 / 1",
+          }}
+          onDragEnd={handleDragEnd}
+          transition={{ type: "spring", duration: 0.5 }}
+        >
```
Unfortunately I was not able to transfer all `styles` into the styled component. The dynamic values such as x, rotate and opacity cannot be processed directly in the styled components as they come from Framer Motion and are animated at runtime. I tried to pass it on as props but that didn't work.

IsabellUrban commented:
Thank you Andrea for your detailed feedback. I have implemented your suggested comments and in some cases added a comment myself. I'm happy to finally finish this Feature.
